src/common/ColourPalette.cpp

```diff
@@ -38,10 +50,342 @@ uInt32 ColourPalette::getRGB(int val) const {
     return m_palette[val];
 }
 
+const uInt32 *ColourPalette::getPalette()
+{
+    return m_palette;
+}
+
+void ColourPalette::setPalette(const string& type,
+                               const string& displayFormat)
+{
+    // See which format we should be using
+    int paletteNum = 0;
+    if(type == "standard")
+        paletteNum = 0;
+    else if(type == "z26")
+        paletteNum = 1;
+    else if(type == "user" && myUserPaletteDefined)
+        paletteNum = 2;
+
+    int paletteFormat = 0;
+    if (displayFormat.compare(0, 3, "PAL") == 0)
+        paletteFormat = 1;
+    else if (displayFormat.compare(0, 5, "SECAM") == 0)
+        paletteFormat = 2;
+
+    m_palette  = availablePalettes[paletteNum][paletteFormat];
+}
+
+void ColourPalette::calculateGrayscaleValues()
+{
+    for (int type = 0; type < 3; type++)
```

Comment:
Well, ALE in general is not thread safe. Checking does not help much unless there is some proper synchronisation mechanism in place... Unless we move to C++ 11 (which I understood you prefer not to), there is no way of making this really thread safe. In addition, every time you load a rom, you need to revise this assignment. If we set m_palette only once, when you load a second rom, you might get incorrect results.

Regarding the grayscale values, yes it could be possible to calculate them once except for the user palettes. I wanted to keep things backwards compatible so kept the, the user palette functionality. I don't know you're opinion about that. Nevertheless, it make take just a few more milliseconds to load the rom but the code is much simpler (unless we totally delete user palettes).